### PR TITLE
Update code coverage step

### PR DESCRIPTION
**Pull request by NotMyFault**
According to the [pipeline docs](https://www.jenkins.io/doc/pipeline/steps/code-coverage-api/#publishcoverage-publish-coverage-report-deprecated), `publishCoverage` is deprecated. 

The PR moves the step over to `recordCoverage`. I chose to store files for the last build only, given our recent disk size issues on ci.j, I don't believe storing all source code files is the most desired approach at this time.

### Testing done

This PR records the coverage.

### Proposed changelog entries

- Entry 1: Issue, human-readable text
- […]

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [ ] The Jira issue, if it exists, is well-described.
- [ ] The changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developers, depending on the change) and are in the imperative mood (see [examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)).
  - Fill in the **Proposed upgrade guidelines** section only if there are breaking changes or changes that may require extra steps from users during upgrade.
- [ ] There is automated testing or an explanation as to why this change has no tests.
- [ ] New public classes, fields, and methods are annotated with `@Restricted` or have `@since TODO` Javadocs, as appropriate.
- [ ] New deprecations are annotated with `@Deprecated(since = "TODO")` or `@Deprecated(forRemoval = true, since = "TODO")`, if applicable.
- [ ] New or substantially changed JavaScript is not defined inline and does not call `eval` to ease future introduction of Content Security Policy (CSP) directives (see [documentation](https://www.jenkins.io/doc/developer/security/csp/)).
- [ ] For dependency updates, there are links to external changelogs and, if possible, full differentials.
- [ ] For new APIs and extension points, there is a link to at least one consumer.

### Desired reviewers

@mention

### Maintainer checklist

Before the changes are marked as `ready-for-merge`:

- [ ] There are at least two (2) approvals for the pull request and no outstanding requests for change.
- [ ] Conversations in the pull request are over, or it is explicit that a reviewer is not blocking the change.
- [ ] Changelog entries in the pull request title and/or **Proposed changelog entries** are accurate, human-readable, and in the imperative mood.
- [ ] Proper changelog labels are set so that the changelog can be generated automatically.
- [ ] If the change needs additional upgrade steps from users, the `upgrade-guide-needed` label is set and there is a **Proposed upgrade guidelines** section in the pull request title (see [example](https://github.com/jenkinsci/jenkins/pull/4387)).
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins.io/issues/?filter=12146)).

<a href="https://gitpod.io/#https://github.com/jenkinsci/jenkins/pull/7831"></a>

### CODE REVIEW

File: Jenkinsfile
Comment (basil) [inline Jenkinsfile:89]: https://github.com/jenkins-infra/pipeline-library/pull/649 uses `MODIFIED` rather than `LAST_BUILD`, so I think we should be consistent with that. If `LAST_BUILD` is preferable to `MODIFIED`, then I think it should be changed in both `pipeline-library` _and_ in this repository.

Comment (NotMyFault) [reply Jenkinsfile]: Oh, I didn't see the PR. Ulli provided a reasonable explanation why he picked 'MODIFIED'.
Based on that, I shifted our retention too.

**Review by basil — APPROVED** [review]

**Review by basil — COMMENTED** [review]
[Build is failing](https://ci.jenkins.io/job/Core/job/jenkins/job/PR-7831/6/execution/node/275/).

**Comment by basil** [comment]
@uhafner Is this a bug in the code coverage plugin?

**Comment by uhafner** [comment]
It seems that there are more coverage reports found that match the same pattern *and* contain the same classes. Is this true? Currently I throw an exception if I get diverging coverage information for the same class.

**Comment by basil** [comment]
I have no idea, I am just an end user trying to upgrade from the old (deprecated) version to the new version, which causes builds to start failing when they were working before.

**Comment by NotMyFault** [comment]
> It seems that there are more coverage reports found that match the same pattern _and_ contain the same classes. Is this true?

If the question is whether there are multiple files in this PR's workspace, I'm afraid I'm unable to check that assumption. Someone else would need to take a look.

**Review by basil — CHANGES_REQUESTED** [review]
Build is failing. You are the author of the PR, so the expectation is that you would do any necessary work to complete the PR.

**Comment by NotMyFault** [comment]
> Build is failing. You are the author of the PR, so the expectation is that you would do any necessary work to complete the PR.

As I said before, I'm unable to check the PRs workspace to verify Ulli's thought.

**Comment by basil** [comment]
I will remove the `needs-fix` label once the build is passing.

**Comment by uhafner** [comment]
I did run the coverage locally now. The step actually reads a single file only. So the configuration is correct, but the aggregation report of Jenkins core is not yet supported by my coverage-model. I will track down the root cause and update the parser.

**Comment by uhafner** [comment]
Released in https://github.com/jenkinsci/code-coverage-api-plugin/releases/tag/v4.3.0

Once this version is deployed on ci.jenkins.io you can restart the build and test.

**Comment by lemeurherve** [comment]
Build restarted.

**Comment by uhafner** [comment]
Looks good now!

Bildschirmfoto 2023-04-18 um 13 51 54

File: Jenkinsfile
Comment (uhafner) [reply Jenkinsfile]: I used `sourceCodeRetention: 'MODIFIED'`, so we only will see the source code rendering for modified files. Otherwise a lot of unused files will be stored in our workspaces.

**Review by basil — APPROVED** [review]
This PR is now ready for merge. We will merge it after approximately 24 hours if there is no negative feedback. Please see the [merge process documentation](https://github.com/jenkinsci/jenkins/blob/master/docs/MAINTAINERS.adoc#merge-process) for more information about the merge process. Thanks!

**Review by uhafner — APPROVED** [review]

**Pull request merged**
